src/componententity.cpp

```diff
@@ -246,34 +333,83 @@ size_t ComponentEntity::componentCount() const
     return mPimpl->mComponents.size();
 }
 
-bool ComponentEntity::containsComponent(const std::string &name) const
+bool ComponentEntity::containsComponent(const std::string &name, bool searchEncapsulated) const
 {
-    auto result = mPimpl->findComponent(name);
-    return result != mPimpl->mComponents.end();
+    bool result = false;
+    if (containsComponentInThis(name)) {
+        result = true;
+    } else if (searchEncapsulated) {
+        for (size_t i = 0; i < componentCount(); ++i) {
+            result = getComponent(i)->containsComponent(name);
+            if (result) {
+                break;
+            }
+        }
+    }
+    return result;
 }
 
-ComponentPtr ComponentEntity::getComponent(size_t index)
+bool ComponentEntity::containsComponent(const ComponentPtr &component, bool searchEncapsulated) const
 {
-    return mPimpl->mComponents.at(index);
+    bool result = false;
+    if (containsComponentInThis(component)) {
+        result = true;
+    } else if (searchEncapsulated) {
+        for (size_t i = 0; i < componentCount(); ++i) {
+            result = getComponent(i)->containsComponent(component);
+            if (result) {
+                break;
+            }
+        }
+    }
+    return result;
 }
 
-const ComponentPtr& ComponentEntity::getComponent(size_t index) const
+ComponentPtr ComponentEntity::getComponent(size_t index)
 {
     return mPimpl->mComponents.at(index);
 }
 
-ComponentPtr ComponentEntity::getComponent(const std::string &name)
+const ComponentPtr& ComponentEntity::getComponent(size_t index) const
 {
-    auto result = mPimpl->findComponent(name);
-    size_t index = result - mPimpl->mComponents.begin();
     return mPimpl->mComponents.at(index);
 }
 
-const ComponentPtr& ComponentEntity::getComponent(const std::string &name) const
+ComponentPtr ComponentEntity::getComponent(const std::string &name, bool searchEncapsulated)
 {
-    auto result = mPimpl->findComponent(name);
-    size_t index = result - mPimpl->mComponents.begin();
-    return mPimpl->mComponents.at(index);
+    ComponentPtr foundComponent = nullptr;
+    if (containsComponentInThis(name)) {
+        foundComponent = getComponentInThis(name);
+    } else if (searchEncapsulated) {
+        for (size_t i = 0; i < componentCount(); ++i) {
+            foundComponent = getComponent(i)->getComponent(name);
+            if (foundComponent) {
+                break;
+            }
+        }
+    }
+    return foundComponent;
+}
+
+const ComponentPtr& ComponentEntity::getComponent(const std::string &name, bool searchEncapsulated) const
```

Comment:
The issue may be because you're returning a pointer that is const, rather than a pointer to a const component? So when you call getComponent(i) the component that gives isn't const. You may also need to pass ,true to containsComponent & getComponent(name).